[YouTube] Add support for extracting auto-translated captions

TobiGr opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [ ] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

Extract auto-translated captions for YouTube videos.

## API changes :green_circle: 
### `SubtitlesStream`
This adds `isAutoTranslated()` next to `isAutoGenerated()` to distinguish between auto-generated subtitles which use speech2text and auto-translated captions based on Google translator.
Additionally, `getBaseLocale()`, `getDisplayBaseLanguageName()` and `getBaseLanguageTag()` were added to access info on the language which was used for auto-translations.

## Issues closed by this PR

Closes TeamNewPipe/NewPipeExtractor#977
Based on and adresses TeamNewPipe/NewPipe#8023

AudricV commented on extractor/src/main/java/org/schabi/newpipe/extractor/stream/SubtitlesStream.java:
```diff
@@ -150,6 +152,18 @@ public Builder setAutoGenerated(final boolean autoGenerated) {
         }
 
+        /**
+         * Set whether the subtitles have been automatically
+         * (i.e. by a machine like Google Translator) translated by the streaming service.
```
```suggestion
         * Set whether the subtitles have been automatically translated
         * (i.e. by a machine like Google Translator) by the streaming service.
         *
```

AudricV commented on extractor/src/main/java/org/schabi/newpipe/extractor/stream/SubtitlesStream.java:
```diff
@@ -217,6 +237,7 @@ public SubtitlesStream build() {
      * @param deliveryMethod the {@link DeliveryMethod} of the stream
      * @param languageCode   the language code of the stream
      * @param autoGenerated  whether the subtitles are auto-generated by the streaming service
+     * @param autoTranslated
```
Parameter description is missing.

AudricV commented on extractor/src/main/java/org/schabi/newpipe/extractor/stream/SubtitlesStream.java:
```diff
@@ -276,6 +299,19 @@ public boolean isAutoGenerated() {
 
+    /**
+     * Whether the subtitles are translated automatically by a machine.
+     * <p>
+     * Some streaming services provide automatically translated subtitles.
+     * YouTube, for example, uses Google translator to generate translated subtitles.
+     * Automatically translated subtitles might not coincide completely with the original text.
+     * </p>
+     * @return {code true} if the subtitles are auto-translated, {@link false} otherwise
```
```suggestion
     *
     * <p>
     * Some streaming services provide automatically translated subtitles.
     * YouTube, for example, uses Google Translator to generate translated subtitles.
     * Automatically translated subtitles might not coincide completely with the original text.
     * </p>
     *
     * @return {code true} if the subtitles are auto-translated, {@link false} otherwise
```

AudricV commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java:
```diff
@@ -717,13 +716,32 @@ public List<SubtitlesStream> getSubtitles(final MediaFormat format) throws Parsi
                         .setAutoGenerated(isAutoGenerated)
+                        .setAutoTranslated(false)
                         .build());
+                if (i == 0 && caption.getBoolean("isTranslatable")
```
I would not base the extraction on the index, but rather on whether the subtitles are auto-generated:

```suggestion
                if (isAutoGenerated && caption.getBoolean("isTranslatable")
```

Also, this PR doesn't add support of subtitles translation for uploaded subtitles. For instance, see https://www.youtube.com/watch?v=_cMxraX_5RE: you can translate from German to French and from English to French, and the translations are different.

We may need another property in `SubtitlesStream` for this.

TobiGr replied:
I don't understand why we should use `isAutoGenerated` here. For better quality, it should be `!isAutoGenerated`. Manually added captions should be exact.
I was also wondering whether we should provide the auto-translated captions by default. Extracting the data for and generating ~100 SubtitleStreams takes some time. I'd definitely not recommend to do this for all available languages by default. On the other hand, we could provide a method which does this when needed.

AudricV commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java:
```diff
@@ -717,13 +716,32 @@ public List<SubtitlesStream> getSubtitles(final MediaFormat format) throws Parsi
+                                .build());
+                    }
+                }
+
```
```suggestion
```

TobiGr committed:
[YouTube] Add support for extracting auto-translated captions

Closes TeamNewPipe/NewPipeExtractor#977
Based on and adresses TeamNewPipe/NewPipe#8023

Facni commented:
What happened to this?

TobiGr committed:
[YouTube] Optimize extracting auto-translated captions

Faster and ordered: captions provided by the user are at the beginning of the list, auto-translated captions are at the end

TobiGr committed:
[YouTube] Add tests for auto-translated captions

TobiGr committed:
Add mocks for new tests
